### DISCUSSION TITLE
Allow 'record' as a generic type constraint

### DISCUSSION THREAD

**Discussion by WhitWaldo** [dotnet/csharplang#7106] [Language Ideas]
I'd like to propose that just as `class` can be used as a generic type constraint, `record` be similarly allowable as a constraint precisely so that I can use the `with` keyword in a method in which I'm passing a generically typed record. 

I would ideally like to be able to enable something like the following:

```csharp
public record Person(string Name, bool IsEnabled);

public static class Group<TItem> where TItem : record
{
  public static TItem ReduceDisabledAction(TItem state)
  {
    return state with
    {
      IsEnabled = false
    };
  }
}
```

If this class isn't generic and I pass a record into it, it works fine, but as it stands today, Visual Studio reports that "TItem is not a valid record or value type" because I have no means of indicating that it is. 

As with all use cases for generics, supporting this would save me having to create and maintain several hundred classes of duplicated type-specific classes and allow me to use a single generic set of classes for the same.

Thank you for the consideration.

**Comment by HaloFour** [comment]
This would be enabled by making `with` pattern-based, so that other non-record types could take advantage of it:

#162 
https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-09-26.md#construction-improvements

That would enable you to define an interface that exposes the expected pattern which could be used as the generic type constraint.
